### ISSUE TITLE
address of TEST occasional failure occurs in zpool_import_001_pos where it fails to import by guid

### ISSUE THREAD

**Issue by liaoyuxiangqin** [zfsonlinux/zfs#4987]
Dear All:
    Some address information of TEST occasional failure occurs in zpool_import_001_pos where it fails to import by guid as follows:

conditions:
OS: Centos7 3.10.0-123.el7.x86_64
ZFS: zfs-cf2731e65b2015988b2cae7970886279e11b013f
SPL: spl-576865be20ce4a0d4365cd62a589edec070fe08c

 Failures: From the follow log we can learn that `/dev/loop0` have parted to 4 partitions, and `testpool.32389` create based on `/dev/loop0p1`, and the pool's guid is `4012590715811873734`. Further more, from the import by guid fail result is 'no such pool available' we can learn `the reason may be is that /dev/loop0p1 lost or pool label lost`, so continue to the following tests and analyses.

```
Test: /home/yuxiang/zfs-cf2/tests/zfs-tests/tests/functional/cli_root/zpool_import/setup (run as root) [02:18] [PASS]
11:15:30.36 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zpool/zpool create -f foopool15672 loop0
11:15:30.63 SUCCESS: cleanup_devices loop0
11:15:30.99 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zpool/zpool create -f foopool15672 loop1
11:15:31.26 SUCCESS: cleanup_devices loop1
11:15:31.35 SUCCESS: /usr/sbin/parted /dev/loop0 -s -- mklabel gpt
11:15:31.51 SUCCESS: /usr/sbin/parted /dev/loop0 -s -- mkpart part0 1cyl 16cyl
11:16:54.63 SUCCESS: set_partition 0  128m loop0
11:16:55.53 SUCCESS: /usr/sbin/parted /dev/loop0 -s -- mkpart part1 17cyl 32cyl
11:16:56.18 SUCCESS: set_partition 1 17 128m loop0
11:16:57.36 SUCCESS: /usr/sbin/parted /dev/loop0 -s -- mkpart part2 32cyl 47cyl
11:16:57.99 SUCCESS: set_partition 2 32 128m loop0
11:16:59.51 SUCCESS: /usr/sbin/parted /dev/loop0 -s -- mkpart part3 48cyl 63cyl
11:16:59.94 SUCCESS: set_partition 3 48 128m loop0
11:17:03.30 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zpool/zpool create -f testpool.32389 loop0p1
11:17:04.46 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zfs/zfs create testpool.32389/testfs.32389
11:17:04.55 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zfs/zfs set mountpoint=/var/tmp/testdir32389 testpool.32389/testfs.32389
11:17:04.68 SUCCESS: /usr/sbin/parted /dev/loop1 -s -- mklabel gpt
11:17:04.75 SUCCESS: /usr/sbin/parted /dev/loop1 -s -- mkpart part0 1cyl 128cyl
11:17:43.37 SUCCESS: set_partition 0  1g loop1
11:17:43.49 SUCCESS: /usr/bin/mkdir -p /var/tmp/dev_import-test
11:17:43.50 SUCCESS: /usr/bin/mount /dev/loop1p1 /var/tmp/dev_import-test
11:17:43.88 SUCCESS: /home/yuxiang/zfs-cf2/tests/zfs-tests/cmd/mkfile/mkfile 64m /var/tmp/dev_import-test/disk0
11:17:44.29 SUCCESS: /home/yuxiang/zfs-cf2/tests/zfs-tests/cmd/mkfile/mkfile 64m /var/tmp/dev_import-test/disk1
11:17:44.74 SUCCESS: /home/yuxiang/zfs-cf2/tests/zfs-tests/cmd/mkfile/mkfile 64m /var/tmp/dev_import-test/disk2
11:17:46.37 SUCCESS: /home/yuxiang/zfs-cf2/tests/zfs-tests/cmd/mkfile/mkfile 64m /var/tmp/dev_import-test/disk3
11:17:48.10 SUCCESS: /home/yuxiang/zfs-cf2/tests/zfs-tests/cmd/mkfile/mkfile 64m /var/tmp/dev_import-test/disk4
Test: /home/yuxiang/zfs-cf2/tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_001_pos (run as root) [00:03] [FAIL]
11:17:48.50 ASSERTION: Verify that an exported pool can be imported.
11:17:49.42 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zpool/zpool create -f testpool1.32389 /var/tmp/dev_import-test/disk0 /var/tmp/dev_import-test/disk1 /var/tmp/dev_import-test/disk2
11:17:49.48 SUCCESS: /usr/bin/mkdir -p /var/tmp/testdir132389
11:17:49.55 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zfs/zfs create testpool1.32389/testfs.32389
11:17:49.60 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zfs/zfs set mountpoint=/var/tmp/testdir132389 testpool1.32389/testfs.32389
11:17:51.38 SUCCESS: /usr/bin/cp /home/yuxiang/zfs-cf2/tests/zfs-tests/include/libtest.shlib /var/tmp/testdir32389/testfile0.32389
11:17:51.41 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zfs/zfs umount /var/tmp/testdir32389
11:17:51.45 SUCCESS: /home/yuxiang/zfs-cf2/cmd/zpool/zpool export testpool.32389
11:17:51.76 ERROR: /home/yuxiang/zfs-cf2/cmd/zpool/zpool import 4012590715811873734 exited 1
11:17:51.84 cannot import 'testpool.32389': no such pool available
11:17:51.84 ERROR: /home/yuxiang/zfs-cf2/cmd/zpool/zpool import testpool.32389 exited 1
```
Analysis:   
    step1: from zdb dump we can learn that the pool is export success and it's label is correct, so i think failure reason is '/dev/loop0p1' lost, and continue to the step2.

```
[root@localhost tmp]# zdb -l /dev/loop0p1 
    --------------------------------------------
    LABEL 0
    --------------------------------------------
        version: 5000
        name: 'testpool.32389'
        state: 1
        txg: 13
        pool_guid: 4012590715811873734
        errata: 0
        hostname: 'localhost.localdomain'
        top_guid: 5718357308702798668
        guid: 5718357308702798668
        vdev_children: 1
        vdev_tree:
            type: 'disk'
            id: 0
            guid: 5718357308702798668
            path: '/dev/loop0p1'
            whole_disk: 0
            metaslab_array: 67
            metaslab_shift: 24
            ashift: 9
            asize: 119013376
            is_log: 0
            create_txg: 4
        features_for_read:
            com.delphix:hole_birth
            com.delphix:embedded_data
    --------------------------------------------
    LABEL 1
    --------------------------------------------
        version: 5000
        name: 'testpool.32389'
        state: 1
        txg: 13
        pool_guid: 4012590715811873734
        errata: 0
        hostname: 'localhost.localdomain'
        top_guid: 5718357308702798668
        guid: 5718357308702798668
        vdev_children: 1
        vdev_tree:
            type: 'disk'
            id: 0
            guid: 5718357308702798668
            path: '/dev/loop0p1'
            whole_disk: 0
            metaslab_array: 67
            metaslab_shift: 24
            ashift: 9
            asize: 119013376
            is_log: 0
            create_txg: 4
        features_for_read:
            com.delphix:hole_birth
            com.delphix:embedded_data
    --------------------------------------------
    LABEL 2
    --------------------------------------------
        version: 5000
        name: 'testpool.32389'
        state: 1
        txg: 13
        pool_guid: 4012590715811873734
        errata: 0
        hostname: 'localhost.localdomain'
        top_guid: 5718357308702798668
        guid: 5718357308702798668
        vdev_children: 1
        vdev_tree:
            type: 'disk'
            id: 0
            guid: 5718357308702798668
            path: '/dev/loop0p1'
            whole_disk: 0
            metaslab_array: 67
            metaslab_shift: 24
            ashift: 9
            asize: 119013376
            is_log: 0
            create_txg: 4
        features_for_read:
            com.delphix:hole_birth
            com.delphix:embedded_data
    --------------------------------------------
    LABEL 3
    --------------------------------------------
        version: 5000
        name: 'testpool.32389'
        state: 1
        txg: 13
        pool_guid: 4012590715811873734
        errata: 0
        hostname: 'localhost.localdomain'
        top_guid: 5718357308702798668
        guid: 5718357308702798668
        vdev_children: 1
        vdev_tree:
            type: 'disk'
            id: 0
            guid: 5718357308702798668
            path: '/dev/loop0p1'
            whole_disk: 0
            metaslab_array: 67
            metaslab_shift: 24
            ashift: 9
            asize: 119013376
            is_log: 0
            create_txg: 4
        features_for_read:
            com.delphix:hole_birth
            com.delphix:embedded_data
    [root@localhost tmp]# 
```

step2: add some log information in function: zpool_find_import_blkid and manual import pool by guid, 
from follows log we learn that use libblkid enumerate all known zfs devices not search `/dev/loop0p1`, so that we can't read label config from /dev/loop0p1, and return 'no such pool available'. So i think the `blkid lib` in current OS exist bug, and occasional search partition dev exist miss.

```
[root@localhost tmp]# zpool import 4012590715811873734
zpool_do_import:name[(nil)] guid[4012590715811873734] file[(nil)] unique[-1] san[0] path[0] begin>>>>>
zpool_search_import:name[(nil)] guid[4012590715811873734] exist[0] >>>>>
zpool_find_import_blkid:blkid_dev_next rn_name[/dev/loop2]>>>>>
zpool_find_import_blkid:blkid_dev_next rn_name[/dev/loop0]>>>>>
zpool_find_import_blkid:blkid_dev_next rn_name[/dev/loop1]>>>>>
zpool_do_import:2533:cannot import '4012590715811873734': no such pool available
[root@localhost tmp]#
```
step3: use blkid tool to check zfs devices whether exist and correct.

```
[root@localhost ~]# rpm -qa |grep libblkid*
libblkid-devel-2.23.2-16.el7.x86_64
libblkid-2.23.2-16.el7.x86_64
```
scenario1: when import pool by guid success, we searched `/dev/loop0p1` is zfs device.
```
[root@localhost dev]# blkid -t TYPE="zfs_member"   
/dev/loop0: LABEL="foopool32126" UUID="6012303043484070649" UUID_SUB="17530838439788337295" TYPE="zfs_member" 
/dev/loop0p1: LABEL="testpool.32389" UUID="6104980183779630722" UUID_SUB="11512063207612755591" TYPE="zfs_member" PARTLABEL="part0" PARTUUID="c7893d48-6a95-40d3-8dc4-2db29ce6ab3a" 
/dev/loop1: LABEL="foopool32126" UUID="14228737093700064280" UUID_SUB="9895466244927853809" TYPE="zfs_member" 
[root@localhost dev]#
```
scenario2: when import pool by guid fail, we can't search `/dev/loop0p1` as zfs device.
```
[root@localhost dev]# blkid -t TYPE="zfs_member"
/dev/loop0: LABEL="foopool8403" UUID="11985523947629466248" UUID_SUB="8903826087660517904" TYPE="zfs_member" 
/dev/loop1: LABEL="foopool8403" UUID="17591282247125289509" UUID_SUB="545812641500644200" TYPE="zfs_member" 
[root@localhost dev]# 
```    
And then, we download and install latest `e2fsprogs-1.43.1` tools which iclude libblkid, and we also can search `/dev/loop0p1` is zfs device, although the device tag of TYPE's value exist differences, but also can explain the latest blkid lib is working correctly.
```
[root@localhost ~]# blkid 
/dev/sda1: UUID="62e2db02-a6a1-434f-82fe-df6b853ecace" TYPE="xfs" 
/dev/sda2: UUID="WpGMey-Tvld-aSRJ-DnjZ-B7tW-txsk-m5kS1i" TYPE="lvm2pv" 
/dev/sr0: LABEL="CentOS 7 x86_64" TYPE="iso9660" 
/dev/mapper/centos-swap: TYPE="swap" UUID="0c3996b1-15d7-4669-8ac5-6f129caeb47f" 
/dev/mapper/centos-root: UUID="0fc2156e-91a3-49ce-ac33-a0a4a17b0584" TYPE="xfs" 
/dev/loop0: TYPE="zfs" 
/dev/loop0p1: TYPE="zfs" 
/dev/loop1: TYPE="zfs" 
/dev/loop1p1: UUID="85f21f6d-abde-4f56-9f98-2bd481307280" TYPE="ext2" 
[root@localhost ~]#
``` 
Solution:
From what has been discussed above, i think the reason of TEST occasional failure occurs in 
zpool_import_001_pos where it fails to import by guid is blkid lib in centos 7 exist bug, 
but imperfect that i have not been able to update blkid lib to latest in current OS and prove my analysis.  So welcome to everybody offer advice and instruction,thanks!

**Comment by liaoyuxiangqin** [comment]
@behlendorf The address information of TEST occasional failure occurs in zpool_import_001_pos where it fails to import by guid as above described, I found the reason for failure is use libblkid enumerate all known zfs devices but miss '/dev/loop0p1', so we can't read pool label config and return 'no such pool available'.

And then  we download and install latest e2fsprogs-1.43.1 tools which iclude libblkid, and we also can search '/dev/loop0p1' is zfs device, although the device tag of TYPE's value exist differences. 

At last, i think the blkid lib in Centos7 not very good to search loop device partition, but imperfect that i have not been able to update libblkid rpm pkg to latest in Centos7 and prove my analysis. so welcome to offer advice and instruction,thanks!

**Comment by behlendorf** [comment]
@liaoyuxiangqin thanks for digging in to this.  Your analysis looks entirely reasonable and I can definitely believe that the libblkid in the CentOS 7 VM isn't handling this entirely correctly.  But I'm unclear if you're saying updating to the latest version will resolve this issue.  Do you have an exact version of e2fsprogs where the issue was resolved?

Alternately, if we can'y rely on libblkid for this test (which would be unfortunate) you could add the **-s** from to `zpool import`.  This effectively disables use of the libblkid cache for the import and resorts to falling back to directory scanning which should always work.

**Comment by behlendorf** [comment]
@liaoyuxiangqin could you please review the proposed fix in #5047.
